File: bindings/r/R/safe_uuid.R
# SPDX-License-Identifier: AGPL-3.0-or-later
# SPDX-FileCopyrightText: 2025 Hyperpolymath

#' Safe UUID validation, parsing, and generation.
#'
#' Provides an S3 class for UUIDs with parsing, formatting, and version detection.

#' UUID version constants.
#' @export
UUID_VERSION_1 <- 1L
#' @export
UUID_VERSION_3 <- 3L
#' @export
UUID_VERSION_4 <- 4L
#' @export
UUID_VERSION_5 <- 5L

#' Nil UUID constant (all zeros).
#' @export
UUID_NIL <- "00000000-0000-0000-0000-000000000000"

#' Create a UUID object.
#'
#' @param bytes Raw vector of 16 bytes
#' @return UUID S3 object
#' @keywords internal
new_uuid <- function(bytes) {

  if (length(bytes) != 16) {
    stop("UUID must be exactly 16 bytes")
  }
  structure(
    list(bytes = bytes),
    class = "UUID"
  )
}

#' Check if object is a UUID.
#'
#' @param x Object to check
#' @return TRUE if x is a UUID object
#' @export
is_uuid <- function(x) {
  inherits(x, "UUID")
}

#' Parse a UUID string.
#'
#' Parses standard UUID format (8-4-4-4-12 hex digits with hyphens).
#'
#' @param uuid_string UUID string to parse
#' @return UUID object or NULL on invalid input
#' @export
parse_uuid <- function(uuid_string) {
  if (is.na(uuid_string) || is.null(uuid_string)) {
    return(NULL)
  }

  # Remove braces if present (Microsoft format)
  uuid_string <- gsub("^\\{|\\}$", "", uuid_string)

  # Validate format: 8-4-4-4-12 hex digits
  pattern <- "^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$"
  if (!grepl(pattern, uuid_string)) {
    return(NULL)
  }

  # Remove hyphens and convert to bytes
  hex_string <- gsub("-", "", uuid_string)

  bytes <- tryCatch({
    n <- nchar(hex_string)
    raw_bytes <- sapply(seq(1, n, by = 2), function(i) {
      strtoi(substr(hex_string, i, i + 1), base = 16L)
    })
    as.raw(raw_bytes)
  }, error = function(e) NULL)

  if (is.null(bytes) || length(bytes) != 16) {
    return(NULL)
  }

  new_uuid(bytes)
}

#' Format UUID object as string.
#'
#' @param x UUID object
#' @param ... Additional arguments (ignored)
#' @return Formatted UUID string in lowercase
#' @export
format.UUID <- function(x, ...) {
  format_uuid(x)
}

#' Print UUID object.
#'
#' @param x UUID object
#' @param ... Additional arguments (ignored)
#' @export
print.UUID <- function(x, ...) {
  cat("UUID:", format_uuid(x), "\n")
  invisible(x)
}

#' Format a UUID object as a string.
#'
#' @param uuid UUID object
#' @param uppercase Use uppercase hex digits (default FALSE)
#' @return Formatted UUID string
#' @export
format_uuid <- function(uuid, uppercase = FALSE) {
  if (is.null(uuid) || !is_uuid(uuid)) {
    return(NA_character_)
  }

  hex_chars <- format(as.hexmode(as.integer(uuid$bytes)), width = 2)
  hex_string <- paste0(hex_chars, collapse = "")

  # Insert hyphens at positions 8, 12, 16, 20
  formatted <- paste0(
    substr(hex_string, 1, 8), "-",
    substr(hex_string, 9, 12), "-",
    substr(hex_string, 13, 16), "-",
    substr(hex_string, 17, 20), "-",
    substr(hex_string, 21, 32)
  )

  if (uppercase) {
    toupper(formatted)
  } else {
    tolower(formatted)
  }
}

#' Validate a UUID string format.
#'
#' @param uuid_string UUID string to validate
#' @return TRUE if valid UUID format
#' @export
is_valid_uuid <- function(uuid_string) {
  !is.null(parse_uuid(uuid_string))
}

#' Get the version of a UUID.
#'
#' @param uuid UUID object or string
#' @return UUID version (1, 3, 4, 5) or NA if invalid
#' @export
uuid_version <- function(uuid) {
  if (is.character(uuid)) {
    uuid <- parse_uuid(uuid)
  }

  if (is.null(uuid) || !is_uuid(uuid)) {
    return(NA_integer_)
  }

  # Version is in bits 4-7 of byte 7 (index 7, 0-indexed)
  version_byte <- as.integer(uuid$bytes[7])
  bitwAnd(bitwShiftR(version_byte, 4), 0x0F)
}

#' Get the variant of a UUID.
#'
#' @param uuid UUID object or string
#' @return UUID variant string or NA if invalid
#' @export
uuid_variant <- function(uuid) {
  if (is.character(uuid)) {
    uuid <- parse_uuid(uuid)
  }

  if (is.null(uuid) || !is_uuid(uuid)) {
    return(NA_character_)
  }

  # Variant is in the high bits of byte 9 (index 8, 0-indexed)
  variant_byte <- as.integer(uuid$bytes[9])

  if (bitwAnd(variant_byte, 0x80) == 0x00) {
    "NCS"
  } else if (bitwAnd(variant_byte, 0xC0) == 0x80) {
    "RFC4122"
  } else if (bitwAnd(variant_byte, 0xE0) == 0xC0) {
    "Microsoft"
  } else {
    "Future"
  }
}

#' Check if UUID is the nil UUID.
#'
#' @param uuid UUID object or string
#' @return TRUE if nil UUID
#' @export
is_nil_uuid <- function(uuid) {
  if (is.character(uuid)) {
    uuid <- parse_uuid(uuid)
  }

  if (is.null(uuid) || !is_uuid(uuid)) {
    return(FALSE)
  }

  all(uuid$bytes == as.raw(0))
}

#' Generate a random UUID (version 4).
#'
#' Uses cryptographically secure random bytes from openssl.
#'
#' @return UUID object
#' @export
generate_uuid <- function() {
  # Generate 16 random bytes
  bytes <- openssl::rand_bytes(16)

  # Set version to 4 (random): byte 7, bits 4-7
  bytes[7] <- as.raw(bitwOr(bitwAnd(as.integer(bytes[7]), 0x0F), 0x40))

  # Set variant to RFC4122: byte 9, bits 6-7

  bytes[9] <- as.raw(bitwOr(bitwAnd(as.integer(bytes[9]), 0x3F), 0x80))

  new_uuid(bytes)
}

#' Generate a UUID v4 as a string.
#'
#' @return UUID string
#' @export
generate_uuid_string <- function() {
  format_uuid(generate_uuid())
}

#' Create a nil UUID.
#'
#' @return UUID object with all zero bytes
#' @export
nil_uuid <- function() {
  new_uuid(as.raw(rep(0, 16)))
}

#' Compare two UUIDs for equality.
#'
#' @param uuid1 First UUID (object or string)
#' @param uuid2 Second UUID (object or string)
#' @return TRUE if equal
#' @export
uuids_equal <- function(uuid1, uuid2) {
  if (is.character(uuid1)) {
    uuid1 <- parse_uuid(uuid1)
  }
  if (is.character(uuid2)) {
    uuid2 <- parse_uuid(uuid2)
  }

  if (is.null(uuid1) || is.null(uuid2)) {
    return(FALSE)
  }

  all(uuid1$bytes == uuid2$bytes)
}

#' Equality operator for UUID objects.
#'
#' @param e1 First UUID
#' @param e2 Second UUID
#' @return TRUE if equal
#' @export
`==.UUID` <- function(e1, e2) {
  uuids_equal(e1, e2)
}

#' Get raw bytes of a UUID.
#'
#' @param uuid UUID object or string
#' @return Raw vector of 16 bytes or NULL
#' @export
uuid_bytes <- function(uuid) {
  if (is.character(uuid)) {
    uuid <- parse_uuid(uuid)
  }

  if (is.null(uuid) || !is_uuid(uuid)) {
    return(NULL)
  }

  uuid$bytes
}

#' Create UUID from raw bytes.
#'
#' @param bytes Raw vector of 16 bytes
#' @return UUID object or NULL if invalid
#' @export
uuid_from_bytes <- function(bytes) {
  if (!is.raw(bytes) || length(bytes) != 16) {
    return(NULL)
  }

  new_uuid(bytes)
}

#' Parse UUID without hyphens.
#'
#' @param hex_string 32-character hex string
#' @return UUID object or NULL on invalid input
#' @export
parse_uuid_hex <- function(hex_string) {
  if (is.na(hex_string) || nchar(hex_string) != 32) {
    return(NULL)
  }

  if (!grepl("^[0-9a-fA-F]{32}$", hex_string)) {
    return(NULL)
  }

  # Insert hyphens and parse normally
  formatted <- paste0(
    substr(hex_string, 1, 8), "-",
    substr(hex_string, 9, 12), "-",
    substr(hex_string, 13, 16), "-",
    substr(hex_string, 17, 20), "-",
    substr(hex_string, 21, 32)
  )

  parse_uuid(formatted)
}

#' Format UUID without hyphens.
#'
#' @param uuid UUID object or string
#' @return 32-character hex string or NA
#' @export
format_uuid_hex <- function(uuid) {
  if (is.character(uuid)) {
    uuid <- parse_uuid(uuid)
  }

  if (is.null(uuid) || !is_uuid(uuid)) {
    return(NA_character_)
  }

  hex_chars <- format(as.hexmode(as.integer(uuid$bytes)), width = 2)
  tolower(paste0(hex_chars, collapse = ""))
}

File: bindings/ocaml/lib/safe_uuid.ml
(* SPDX-License-Identifier: PMPL-1.0-or-later *)
(* SPDX-FileCopyrightText: 2025 Hyperpolymath *)

(** Safe UUID validation, parsing, and generation. *)

(** UUID version types. *)
type version =
  | V1  (** Time-based UUID *)
  | V2  (** DCE Security UUID *)
  | V3  (** MD5 hash namespace UUID *)
  | V4  (** Random UUID *)
  | V5  (** SHA-1 hash namespace UUID *)
  | V6  (** Reordered time-based UUID *)
  | V7  (** Unix epoch time-based UUID *)
  | V8  (** Custom UUID *)
  | Unknown of int

(** UUID variant types per RFC 4122. *)
type variant =
  | Ncs       (** Reserved for NCS backward compatibility *)
  | Rfc4122   (** Standard UUID variant *)
  | Microsoft (** Reserved for Microsoft backward compatibility *)
  | Future    (** Reserved for future definition *)

(** Opaque UUID type. *)
type t = {
  bytes: string;  (** 16 bytes raw representation *)
}

type error =
  | Invalid_format
  | Invalid_length
  | Invalid_hex_char

(** Parse a hex character to its integer value. *)
let hex_char_to_int c =
  match c with
  | '0'..'9' -> Ok (Char.code c - Char.code '0')
  | 'a'..'f' -> Ok (Char.code c - Char.code 'a' + 10)
  | 'A'..'F' -> Ok (Char.code c - Char.code 'A' + 10)
  | _ -> Error Invalid_hex_char

(** Convert two hex characters to a byte. *)
let hex_pair_to_byte h l =
  match hex_char_to_int h, hex_char_to_int l with
  | Ok hv, Ok lv -> Ok (Char.chr ((hv lsl 4) lor lv))
  | Error e, _ | _, Error e -> Error e

(** Parse a UUID from standard 8-4-4-4-12 format. *)
let parse str =
  let str = String.trim str in
  (* Remove optional braces/parens *)
  let str =
    if String.length str >= 2 then
      match str.[0], str.[String.length str - 1] with
      | '{', '}' | '(', ')' -> String.sub str 1 (String.length str - 2)
      | _ -> str
    else str
  in
  if String.length str <> 36 then Error Invalid_length
  else if str.[8] <> '-' || str.[13] <> '-' || str.[18] <> '-' || str.[23] <> '-' then
    Error Invalid_format
  else
    let hex_chars =
      String.sub str 0 8 ^
      String.sub str 9 4 ^
      String.sub str 14 4 ^
      String.sub str 19 4 ^
      String.sub str 24 12
    in
    if String.length hex_chars <> 32 then Error Invalid_format
    else
      let buf = Buffer.create 16 in
      let rec convert i =
        if i >= 32 then Ok { bytes = Buffer.contents buf }
        else
          match hex_pair_to_byte hex_chars.[i] hex_chars.[i + 1] with
          | Ok c -> Buffer.add_char buf c; convert (i + 2)
          | Error e -> Error e
      in
      convert 0

(** Format a UUID to standard 8-4-4-4-12 lowercase format. *)
let format uuid =
  let hex_of_byte b =
    Printf.sprintf "%02x" (Char.code b)
  in
  let b = uuid.bytes in
  Printf.sprintf "%s%s%s%s-%s%s-%s%s-%s%s-%s%s%s%s%s%s"
    (hex_of_byte b.[0]) (hex_of_byte b.[1]) (hex_of_byte b.[2]) (hex_of_byte b.[3])
    (hex_of_byte b.[4]) (hex_of_byte b.[5])
    (hex_of_byte b.[6]) (hex_of_byte b.[7])
    (hex_of_byte b.[8]) (hex_of_byte b.[9])
    (hex_of_byte b.[10]) (hex_of_byte b.[11]) (hex_of_byte b.[12])
    (hex_of_byte b.[13]) (hex_of_byte b.[14]) (hex_of_byte b.[15])

(** Format a UUID to uppercase. *)
let format_upper uuid =
  String.uppercase_ascii (format uuid)

(** Get the UUID version. *)
let get_version uuid =
  let version_byte = Char.code uuid.bytes.[6] in
  match (version_byte lsr 4) land 0x0F with
  | 1 -> V1
  | 2 -> V2
  | 3 -> V3
  | 4 -> V4
  | 5 -> V5
  | 6 -> V6
  | 7 -> V7
  | 8 -> V8
  | n -> Unknown n

(** Get the UUID variant. *)
let get_variant uuid =
  let variant_byte = Char.code uuid.bytes.[8] in
  if (variant_byte land 0x80) = 0 then Ncs
  else if (variant_byte land 0xC0) = 0x80 then Rfc4122
  else if (variant_byte land 0xE0) = 0xC0 then Microsoft
  else Future

(** Check if a string is a valid UUID format. *)
let is_valid str =
  match parse str with
  | Ok _ -> true
  | Error _ -> false

(** Generate a random (v4) UUID. *)
let generate_v4 () =
  let buf = Bytes.create 16 in
  let ic = open_in_bin "/dev/urandom" in
  really_input ic buf 0 16;
  close_in ic;
  (* Set version to 4 *)
  Bytes.set buf 6 (Char.chr ((Char.code (Bytes.get buf 6) land 0x0F) lor 0x40));
  (* Set variant to RFC 4122 *)
  Bytes.set buf 8 (Char.chr ((Char.code (Bytes.get buf 8) land 0x3F) lor 0x80));
  { bytes = Bytes.to_string buf }

(** The nil UUID (all zeros). *)
let nil = { bytes = String.make 16 '\000' }

(** Check if a UUID is nil. *)
let is_nil uuid =
  uuid.bytes = nil.bytes

(** Compare two UUIDs. *)
let compare a b =
  String.compare a.bytes b.bytes

(** Check equality of two UUIDs. *)
let equal a b =
  a.bytes = b.bytes

(** Get raw bytes of UUID. *)
let to_bytes uuid = uuid.bytes

(** Create UUID from raw bytes. *)
let of_bytes bytes =
  if String.length bytes <> 16 then Error Invalid_length
  else Ok { bytes }
